docs/OBJECT-LEVEL-ENCRYPTION.md

```diff
@@ -0,0 +1,119 @@
+# OBJECT LEVEL ENCRYPTION
+
+## Motivation
+
+Independent of the use of TLS in the [API](../api/APIv2.md) and [object signing](../api/OBJECT-SIGNING.md) there are some secrets in the device configuration API (such as datastore credentials, WiFi credentials, and cloud-init user data) which should have minimal exposure by
+
+1. keeping them in some vault inside a controller implementation
+2. not exposing them to any microservices in a controller except for the party which decrypts from the vault and re-encrypts for a particular device
+3. not exposing them to all microservices on the device and accidentally storing them in files or putting them in logs
+4. potentially restricting it by policy on the device so that in a defense-in-depth implementation only the downloader can access datastore credentials, only nim can access WiFi credentials (basically least privilege approach)
+5. MiTM proxies and server-side load balancers which terminate TLS should not be able to see this information
+
+## Overview
+
+The approach is to encrypt such credential and other private information using a per-device key, and doing this encryption before assigning it to the fields which get protobuf encoded. Thus other microservices in a controller and the device will pass them along as any protobuf to golang data structure, but the content will be encrypted.
```

Comment:
This is really helpful. From the beginning it explains that there is "usual data" - we always want to protect (encrypt) and validate (sign) that - and "special data", which has to be protected not just in transit with everything else, or even at rest, but accessible only to select services, and should be kept in hardware or software vaults where feasible. I read this, and I get it from the beginning.

Reply:
Maybe I should add 'encrypted at rest' to the list of requirements?
Note that currently in EVE the protobuf messages are saved in /persist/checkpoint, and that is not an encrypted directory.

Reply:
> One part really still wasn't clear. In the case of mitm, if cert exchange happens over the TLS network channel, and the mitm can read everything, and therefore change everything, what prevents mitm from inserting its own ECDH key instead? Or is it the signing mechanism, where the message is signed by a cert whose CA is owned by the controller, and is pre-embedded in the device?

Because the controller cert needs to be signed by the root CA the device trusts for object security. I'll make this clear.